service/raft/raft_group0.cc

```diff
@@ -230,135 +236,228 @@ future<> raft_group0::abort() {
+                    .info = my_info,
+                }}, {});
             break;
+        } catch (abort_requested_exception&) {
+            throw;
         } catch (std::runtime_error& e) {
             // Retry
             rslog.error("failed to modify config at peer {}: {}", g0_info.addr.id, e);
         }
+
         // Try again after a pause
         co_await seastar::sleep_abortable(std::chrono::milliseconds{1000}, _abort_source);
     }
     co_await db::system_keyspace::set_raft_group0_id(group0_id.id);
     // Allow peer_exchange() RPC to access group 0 only after group0_id is persisted.
 
     _group0 = group0_id;
-    rslog.info("{} joined group 0 with id {}", my_addr.id, group0_id);
+    rslog.info("{} joined group 0 with id {}", my_id, group0_id);
 }
 
-future<> raft_group0::become_voter() {
+future<> raft_group0::bootstrap(const std::unordered_set<gms::inet_address>& initial_contact_nodes) {
+    assert(this_shard_id() == 0);
+
     if (!_raft_gr.is_enabled()) {
+        rslog.info("raft_group0::bootstrap(): local RAFT feature disabled, skipping group 0 bootstrap.");
+        // Note: if the local feature was enabled by every node earlier, that would enable the cluster
+        // SUPPORTS_RAFT feature, and the node should then refuse to start during feature check
+        // (because if the local feature is disabled, then the cluster feature - enabled in the cluster - is 'unknown' to us).
         co_return;
     }
-    auto my_addr = co_await load_or_create_my_addr();
-    assert(std::holds_alternative<raft::group_id>(_group0));
-    auto& gid = std::get<raft::group_id>(_group0);
-    if (!_raft_gr.get_server(gid).get_configuration().can_vote(my_addr.id)) {
-        std::vector<raft::server_address> add_set;
-        my_addr.can_vote = true;
-        add_set.push_back(my_addr);
-        auto pause_shutdown = _shutdown_gate.hold();
-        co_return co_await _raft_gr.group0().modify_config(add_set, {});
+
+    if (_client.sys_ks().bootstrap_complete()) {
+        auto group0_id = raft::group_id{co_await db::system_keyspace::get_raft_group0_id()};
+        if (group0_id != raft::group_id{}) {
+            // Group 0 ID is present => we've already joined group 0 earlier.
+            rslog.info("raft_group0::bootstrap(): group 0 ID present. Starting existing Raft server.");
+            co_await start_existing_group0_server(group0_id);
+            co_return;
+        }
+
+        // Scylla has bootstrapped earlier but group 0 ID not present. This means we're upgrading.
     }
+
+    std::vector<raft::server_info> initial_contacts_as_raft_addrs;
+    for (auto& addr: initial_contact_nodes) {
+        if (addr != utils::fb_utilities::get_broadcast_address()) {
+            initial_contacts_as_raft_addrs.push_back(inet_addr_to_raft_addr(addr));
+        }
+    }
+
+    rslog.info("raft_group0::bootstrap(): joining group 0...");
+    co_await join_group0(std::move(initial_contacts_as_raft_addrs), false /* non-voter */);
```

Comment:
May be we should have introduced `raft::voter` while doing `can_vote` refactoring :)